File: cphysics/src/cpp/mass_spring_solver.cpp
/* -*-c++-*- 
 *------------------------------------------------------------------------------
 *                                                                              
 *        openalea.container: utils package                                     
 *                                                                              
 *        Copyright 2006 INRIA - CIRAD - INRA                      
 *                                                                              
 *        File author(s): Jerome Chopard <jerome.chopard@sophia.inria.fr>         
 *                                                                              
 *        Distributed under the Cecill-C License.                               
 *        See accompanying file LICENSE.txt or copy at                          
 *            http://www.cecill.info/licences/Licence_CeCILL-C_V1-en.html       
 *                                                                              
 *        OpenAlea WebSite : http://openalea.gforge.inria.fr                    
 *       
 *                                                                       
 *-----------------------------------------------------------------------------*/

#include <iostream>

#include <algorithm>
#include "cphysics/mass_spring_solver.h"

namespace physics {
namespace mechanics {
	/* *******************************************************
	*
	*	MSSolver2D
	*
	*********************************************************/
	MSSolver2D::~MSSolver2D () {
		/*for (spring_list::iterator it=_spring_list.begin();it!=_spring_list.end();++it) {
			delete *it;
		}*/
		/*for (P2D_list::iterator it=_particules.begin();it!=_particules.end();++it) {
			delete *it;
		}*/
	}

	int MSSolver2D::add_particule () {
		_particules.push_back(Particule2D());
		return _particules.size() - 1;
	}
	
	void MSSolver2D::compute_forces (P2D_list& particules) {
		//set forces to zero
		for (P2D_list::iterator it = particules.begin();
		                        it != particules.end();
		                        ++it) {
			it->force = MSSolver2D::Vector();
		}
		
		//assign forces
		for (spring_list::iterator it = _spring_list.begin();
		                           it != _spring_list.end();
		                           ++it) {
			(*it)->assign_forces(particules);
		}
		
		//boundary condition
		boundary();
	}
	
	void MSSolver2D::update_positions (P2D_list& particules, double dt) {
		for (P2D_list::iterator it = particules.begin();
		                        it != particules.end();
		                        ++it) {
			it->pos += it->force * (dt / it->weight);
			if (it->pos.x() > LARGE or it->pos.y() > LARGE) {
				throw DivergenceError();
			}
		}		
	}
	
	double MSSolver2D::force_max (const P2D_list& particules) {
		double Fmax = 0;
		for (P2D_list::const_iterator it = particules.begin();
		                        it != particules.end();
		                        ++it) {
			Fmax = std::max(Fmax,norm( it->force ) );
		}		
		return Fmax;
	}
	
	double MSSolver2D::deform (P2D_list& particules, double dt) {
		compute_forces(particules);
		update_positions(particules,dt);
		return dt;
	}
	
	void MSSolver2D::deform_to_equilibrium (double dt, double force_threshold) {
		double Fmax = 2 * force_threshold;
		while ( Fmax > force_threshold ) {
			dt = deform(_particules,dt);
			Fmax = force_max(_particules);
			std::cout << "Fmax " << Fmax << std::endl;
		}
	}

	/* *******************************************************
	*
	*	MSSolver3D
	*
	*********************************************************/
	MSSolver3D::~MSSolver3D () {
		/*for (spring_list::iterator it=_spring_list.begin();it!=_spring_list.end();++it) {
			delete *it;
		}*/
		/*for (P3D_list::iterator it=_particules.begin();it!=_particules.end();++it) {
			delete *it;
		}*/
	}

	int MSSolver3D::add_particule () {
		_particules.push_back(Particule3D() );
		return _particules.size() - 1;
	}
	
	void MSSolver3D::compute_forces (P3D_list& particules) {
		//std::cout << "compute" << std::endl;
		//set forces to zero
		for (P3D_list::iterator it = particules.begin();
		                        it != particules.end();
		                        ++it) {
			it->force = MSSolver3D::Vector();
		}
		int i;
		/*PRINT
		i = 0;
		for (P3D_list::iterator it = particules.begin();
		                        it != particules.end();
		                        ++it) {
			std::cout << "F0" << i << " " << it->force << std::endl;
			++i;
		}*/
		
		//assign forces
		for (spring_list::iterator it = _spring_list.begin();
		                           it != _spring_list.end();
		                           ++it) {
			(*it)->assign_forces(particules);
		}
		/*PRINT
		i = 0;
		for (P3D_list::iterator it = particules.begin();
		                        it != particules.end();
		                        ++it) {
			std::cout << "F" << i << " " << it->force << std::endl;
			++i;
		}*/
		
		
		//boundary condition
		cur_part = &particules;
		boundary();
		/*PRINT
		i = 0;
		for (P3D_list::iterator it = particules.begin();
		                        it != particules.end();
		                        ++it) {
			std::cout << "Fb" << i << " " << it->force << std::endl;
			++i;
		}*/
		
	}
	
	void MSSolver3D::update_positions (P3D_list& particules, double dt) {
		for (P3D_list::iterator it = particules.begin();
		                        it != particules.end();
		                        ++it) {
			it->pos += it->force * (dt / it->weight);
			if (it->pos.x() > LARGE or it->pos.y() > LARGE or it->pos.z() > LARGE) {
				throw DivergenceError();
			}
		}		
	}
	
	double MSSolver3D::force_max (const P3D_list& particules) {
		double Fmax = 0;
		for (P3D_list::const_iterator it = particules.begin();
		                        it != particules.end();
		                        ++it) {
			Fmax = std::max(Fmax,norm( it->force ) );
		}		
		return Fmax;
	}
	
	double MSSolver3D::deform (P3D_list& particules, double dt) {
		//std::cout << "MSSolver" << std::endl;
		compute_forces(particules);
		update_positions(particules,dt);
		return dt;
	}
	
	void MSSolver3D::deform_to_equilibrium (double dt, double force_threshold) {
		double Fmax = 2 * force_threshold;
		while ( Fmax > force_threshold ) {
			dt = deform(_particules,dt);
			Fmax = force_max(_particules);
			std::cout << "Fmax " << Fmax << std::endl;
		}
		cur_part = &_particules;
	}

	/* *******************************************************
	*
	*	ForwardEuler
	*
	*********************************************************/
	double ForwardEuler2D::deform (P2D_list& particules, double dt) {
		for (int i = 0; i < 100; ++i) {
			compute_forces(particules);
			update_positions(particules,dt);
		}
		return dt;
	}

	double ForwardEuler3D::deform (P3D_list& particules, double dt) {
		std::cout << "Euler " << dt << std::endl;
		for (int i = 0; i < nb_iterations; ++i) {
			compute_forces(particules);
			update_positions(particules,dt);
			//std::cout << "Fmax " << force_max(particules) << std::endl;
		}
		return dt;
	}

	/* *******************************************************
	*
	*	ForwardMarching
	*
	*********************************************************/
	double ForwardMarching2D::deform (P2D_list& particules, double dt) {
		compute_forces(particules);
		return _deform(particules,dt);
	}
	
	double ForwardMarching2D::_deform (P2D_list& particules, double dt) {
		//compute ref
		P2D_list p_ref(particules);
		update_positions(p_ref,dt);
		
		//compute test
		double dt_test = dt / NB_SUBDI;
		P2D_list p_test(particules);
		update_positions(p_test,dt_test);
		
		for (int i = 0; i < NB_SUBDI; ++i) {
			compute_forces(p_test);
			update_positions(p_test,dt_test);
		}
		
		//put forces in particules
		P2D_list::iterator it = particules.begin();
		P2D_list::iterator it_test = p_test.begin();
		while (it != particules.end() and it_test != p_test.end() ) {
			it->force = it_test->force;
			++ it;
			++ it_test;
		}
		
		//compare
		double err = 0.;
		P2D_list::iterator it_ref = p_ref.begin();
		it_test = p_test.begin();
		while (it_ref != p_ref.end() and it_test != p_test.end() ) {
			err = std::max(err,norm( it_ref->pos - it_test->pos ) );
			++ it_ref;
			++ it_test;
		}
		std::cout << "err" << err << "," << dt << std::endl;
		if (err > ERR_MAX_THRESHOLD) {
			return _deform(particules,dt / 2.);
		}
		else {
		if (err < ERR_MIN_THRESHOLD) {
			return std::min(DT_MAX,dt * 1.3);
		}
		else {
			return dt;
		} }
	}

	double ForwardMarching3D::deform (P3D_list& particules, double dt) {
		//std::cout << "deform" << std::endl;
		compute_forces(particules);
		return _deform(particules,dt);
	}
	
	double ForwardMarching3D::_deform (P3D_list& particules, double dt) {
		//std::cout << "_deform" << std::endl;
		//compute ref
		//std::cout << "ref" << std::endl;
		try {
			P3D_list p_ref(particules);
			update_positions(p_ref,dt);
			
			//compute test
			//std::cout << "test" << std::endl;
			double dt_test = dt / NB_SUBDI;
			P3D_list p_test(particules);
			update_positions(p_test,dt_test);
			
			for (int i = 0; i < NB_SUBDI; ++i) {
				//std::cout << "test" << i << std::endl;
				compute_forces(p_test);
				update_positions(p_test,dt_test);
			}
			/*PRINT
			int i = 0;
			std::cout << "pos";
			for (P3D_list::iterator it = p_test.begin();
			                        it != p_test.end();
			                        ++it) {
				std::cout << " " << it->pos;
				++i;
			}
			std::cout << std::endl;*/
		
			//put forces in particules
			P3D_list::iterator it = particules.begin();
			P3D_list::iterator it_test = p_test.begin();
			while (it != particules.end() and it_test != p_test.end() ) {
				it->force = it_test->force;
				++ it;
				++ it_test;
			}
			
			//compare
			double err = 0.;
			P3D_list::iterator it_ref = p_ref.begin();
			it_test = p_test.begin();
			while (it_ref != p_ref.end() and it_test != p_test.end() ) {
				err = std::max(err,norm( it_ref->pos - it_test->pos ) );
				++ it_ref;
				++ it_test;
			}
			std::cout << "                     err " << err << "," << prev_error << "," << dt << "," << nb_fine_iter << std::endl;
			if (prev_error > 0) {
				//previous step dt *= 1.5
				if (err > (prev_error * 1.4) and err > ERR_MIN_THRESHOLD) {
					std::cout << "                     go back" << std::endl;
					//go back
					nb_fine_iter = 0;
					return _deform(particules,dt / 1.5);
				}
			}
			if (err > ERR_MAX_THRESHOLD) {
				//std::cout << "MAX" << std::endl;
				prev_error = -1.;
				nb_fine_iter = 0;
				return _deform(particules,dt / 2.);
			}
			else {
				//put positions in particules
				it = particules.begin();
				it_test = p_test.begin();
				while (it != particules.end() and it_test != p_test.end() ) {
					it->pos = it_test->pos;
					++ it;
					++ it_test;
				}
				if (err < ERR_MIN_THRESHOLD and nb_fine_iter > 10) {
					prev_error = err;
					nb_fine_iter = 0;
					return std::min(DT_MAX,dt * 1.5);
				}
				else {
					prev_error = -1.;
					nb_fine_iter += 1;
					return dt;
				}
			}
		}
		catch (DivergenceError e) {
			std::cout << "nan " << dt << std::endl;
			nb_fine_iter = 0;
			return _deform(particules,dt / 2.);
		}
	}

};//mechanics
};//physics

File: physics/src/physics/mechanics/mass_spring_solver.py
# -*- python -*-
# -*- coding: latin-1 -*-
#
#       MassSpring : physics package
#
#       Copyright or © or Copr. 2006 INRIA - CIRAD - INRA  
#
#       File author(s): Jerome Chopard <jerome.chopard@sophia.inria.fr>
#
#       Distributed under the Cecill-C License.
#       See accompanying file LICENSE.txt or copy at
#           http://www.cecill.info/licences/Licence_CeCILL-C_V1-en.html
# 
#       VPlants WebSite : https://gforge.inria.fr/projects/vplants/
#

__doc__="""
This module provide an interface for deformation of spring systems
"""

__license__= "Cecill-C"
__revision__=" $Id: mass_spring_solver.py 7882 2010-02-08 18:36:38Z cokelaer $ "

from openalea.plantgl.math import Vector2,Vector3,norm

LARGE = 1e9

class DivergenceError (Exception) :
	pass

class MSSolver (object) :
	"""Integration method for mass spring systems.
	
	basic implementation of mass spring resolution
	using a forward Euler method
	"""
	
	Vector = None
	"""type of vector objects used.
	"""
	
	def __init__ (self, point_weight, spring_list, external_constraint) :
		"""Initialise integrator.
		
		point_weight: a dict of (pid|float)
		spring_list: a list of spring objects
		external_constraint: a function taking this solver as argument
		"""
		self._weight = point_weight
		self._spring_list = spring_list
		self._external_constraint = external_constraint
		self._pos = None
		self._force = None
	
	########################################
	#
	#		accessors
	#
	########################################
	def position (self, pid) :
		"""Retrieve position of a given point.
		
		pid: point id
		return: Vector
		"""
		return self._pos[pid]
	
	def set_position (self, pid, vec) :
		"""Set the position of a given point.
		
		pid: point id
		vec: Vector
		"""
		self._pos[pid] = vec
	
	def force (self, pid) :
		"""Retrieve the force applied on a point
		
		pid: point id
		return: Vector
		"""
		return self._force[pid]
	
	def set_force (self, pid, F) :
		"""Apply a force on a given point.
		
		pid: point id
		F: Vector
		"""
		self._force[pid] = F
	########################################
	#
	#		compute deformation
	#
	########################################
	def compute_forces (self, pos) :
		"""Compute forces acting on points.
		"""
		#compute forces
		force = dict( (pid,self.Vector()) for pid in pos )
		for spring in self._spring_list :
			spring.assign_forces(force,pos)
		
		#external constraints
		self._pos = pos
		self._force = force
		self._external_constraint(self)
		
		#return
		return force
	
	def update_positions (self, pos, force, dt) :
		#compute new positions
		weight = self._weight
		for pid,F in force.iteritems() :
			vec = pos[pid] + F * (dt / weight[pid])
			pos[pid] = vec
			for coord in vec :
				if coord > LARGE :
					raise DivergenceError
	
	def deform (self, pos, dt) :
		"""Compute deformation.
		"""
		#compute forces
		force = self.compute_forces(pos)
		#compute new positions
		self.update_positions(pos,force,dt)
		
		return pos,dt,force
	
	def force_max (self, forces) :
		"""Return biggest force applied on the system.
		"""
		return max(norm(F) for F in forces.itervalues())
	
	def deform_to_equilibrium (self, pos, dt, force_threshold) :
		"""Compute equilibrium positions of the system.
		
		while forces are bigger than a given threshold,
		call deform
		"""
		Fmax = 2 * force_threshold
		while Fmax > force_threshold :
			dt,forces = self.deform(pos,dt)
			Fmax = self.force_max(forces)
			print "Fmax",Fmax


########################################################################
#
#	simple forward euler
#
########################################################################
class ForwardEuler (MSSolver) :
	"""Compute equilibrium using a simple Euler integration scheme.
	
	call deform until forces drop below a given threshold
	"""
	def deform (self, pos, dt, nb_steps = 100) :
		"""Compute nb_steps of deformation.
		"""
		for i in xrange(nb_steps) :
			#compute forces
			force = self.compute_forces(pos)
			#compute new positions
			self.update_positions(pos,force,dt)
		
		return dt,force


class ForwardEuler2D (ForwardEuler) :
	Vector = Vector2


class ForwardEuler3D (ForwardEuler) :
	Vector = Vector3

########################################################################
#
#	Runge et Kuta order 4
#
########################################################################
class RungeKuta (MSSolver) :
	"""Compute equilibrium using a Runge et Kuta integration scheme.
	"""
	
	def deform (self, pos, dt) :
		"""Deform positions using RG4 scheme.
		
		call compute_force for 0, dt /2, dt /2 and dt
		"""
		#F1
		force1 = self.compute_forces(pos)
		#F2
		pos2 = dict( (pid,self.Vector(vec)) for pid,vec in pos.iteritems() )
		self.update_positions(pos2,force1,dt / 2.)
		force2 = self.compute_forces(pos2)
		#F3
		pos3 = dict( (pid,self.Vector(vec)) for pid,vec in pos.iteritems() )
		self.update_positions(pos3,force2,dt / 2.)
		force3 = self.compute_forces(pos3)
		#F4
		pos4 = dict( (pid,self.Vector(vec)) for pid,vec in pos.iteritems() )
		self.update_positions(pos3,force3,dt)
		force4 = self.compute_forces(pos4)
		#compute final displacement
		force = dict( (pid,(force1[pid] +
		                    force2[pid] * 2. + 
		                    force3[pid] * 2. + 
		                    force4[pid]) / 6.) for pid in pos )
		self.update_positions(pos,force,dt)
		#return
		return dt,force


class RungeKuta2D (RungeKuta) :
	Vector = Vector2

class RungeKuta3D (RungeKuta) :
	Vector = Vector3

########################################################################
#
#	Forward Marching
#
########################################################################
class ForwardMarching (MSSolver) :
	"""Compute equilibrium using a forward marching integration scheme.
	
	At each iteration, compare the displacement obtained
	with the given dt and a smaller dt
	adapt dt accordingly
	"""
	DT_MAX = 10.
	NB_SUBDI = 10
	ERR_MAX_THRESHOLD = 1e-4
	ERR_MIN_THRESHOLD = 1e-5
	
	def deform (self, pos, dt) :
		"""Deform positions using ForwardMarching scheme.
		
		compare result with same computation with
		smaller dt
		"""
		self._prev_err = -1
		self._nb_fine_iter = 0
		force = self.compute_forces(pos)
		return self._deform(pos,dt,force)
	
	def _deform (self, pos, dt, force) :
		"""Actually perform the deformation.
		"""
		force_ref = force
		try :
			#compute ref
			pos_ref = dict( (pid,self.Vector(vec)) for pid,vec in pos.iteritems() )
			self.update_positions(pos_ref,force,dt)
			
			#compute test
			dt_test = dt / self.NB_SUBDI
			pos_test = dict( (pid,self.Vector(vec)) for pid,vec in pos.iteritems() )
			self.update_positions(pos_test,force,dt_test)
			
			for i in xrange(self.NB_SUBDI - 1) :
				force = self.compute_forces(pos_test)
				self.update_positions(pos_test,force,dt_test)
		except DivergenceError :
			print "nan",dt
			return self._deform(pos,dt / 2,force_ref)
		
		#compare
		err = max(norm(pos_test[pid] - pos_ref[pid]) for pid in pos)
		print "                        err",err,self._prev_err,dt,self._nb_fine_iter
		if self._prev_err > 0 :
			if err > (self._prev_err * 1.4) and err > self.ERR_MIN_THRESHOLD :
				print "                        go back"
				self._nb_fine_iter = 0
				return self._deform(pos,dt / 1.5)
		if err > self.ERR_MAX_THRESHOLD :
			self._prev_err = -1
			self._nb_fine_iter = 0
			return self._deform(pos,dt / 2,force_ref)
		else :
			for pid,vec in pos_test.iteritems() :
				pos[pid] = vec
			if err < self.ERR_MIN_THRESHOLD and self._nb_fine_iter > 10 :
				self._prev_err = err
				self._nb_fine_iter = 0
				return min(self.DT_MAX,dt * 1.5),force
			else :
				self._prev_err = -1
				self._nb_fine_iter += 1
				return dt,force
		


class ForwardMarching2D (ForwardMarching) :
	Vector = Vector2

class ForwardMarching3D (ForwardMarching) :
	Vector = Vector3

